### ISSUE TITLE
Intermediate Variable Access for temporary solver states

### ISSUE THREAD

**Issue by robbr48** [modelica/fmi-standard#546]
The following suggested changes are required for, but not limited to, 
supporting TLM connections between FMUs.

An FMU should be able to request input variables or expose available 
output variables for the master at any intermediate time during the step. 
This should be achieved by calling the `fmi3IntermediateStepFinished` 
callback function.

The following flags, among others, should be included in the callback.

```
typedef struct{
    fmi3Real intermediateStepTime; 
    [...]
    fmi3Boolean intermediateGetVariablesAllowed;
    fmi3Boolean intermediateSetVariablesAllowed;
    fmi3Boolean internalStepFinished;
    [...]
} fmi3IntermediateStepInfo;`
```
- `intermediateGetVariablesAllowed`: 
If true, the master is allowed to call `getReal` at this intermediate state.
- `intermediateSetVariablesAllowed`: 
If true, the master is allowed to call `setReal` at this intermediate state.
- `internalStepFinisherd`: 
If true, the internal solver in the FMU has completed a successful 
integration step. Hence, outputs variables are valid and can be 
forwarded to other FMUs or written result files. If false, output 
variables are only temporary and should only be used for 
computation of temporary input variables.

**Motivation**

In the TLM case, it is desirable to let the master handle computation 
of effort variables (force,pressure,voltage, etc) from flow variables 
(speed,fluid flow,current etc). This requires the master to read the 
flow variable from the FMU using `getReal` before writing the effort 
variable using `setReal`. If the FMU requires an updated effort 
variable at a temporary solver state, e.g. inside a rhs() function, 
there must be a way to inform the master that the outputs are 
temporary and should not be forwarded anywhere.

**In conclusion** 

If the FMU requires updated inputs at a _temporary solver state_, it 
should call `fmi3IntermediateStepFinished` with the following flags:
- `intermediateGetVariablesAllowed=true`
- `intermediateSetVariablesAllowed=true` 
- `internalStepFinished=false`.

The master should then read intermediate outputs, but only use 
them to compute intermediate inputs.

After a _completed integration step_, the FMU should call the 
`fmi3IntermediateStepFinished` with the following flags: 
- `intermediateGetVariablesAllowed=true`
- `intermediateSetVariablesAllowed=true`
- `internalStepFinished=true`

The master should then read intermediate outputs and forward them 
to other FMUs/result files, and also write updated intermediate inputs.

**Comment by jbernalr** [comment]
are these changes independent of (now closed) issue #534 ?

**Comment by KarlWernersson** [comment]
@jbernalr Yes they are completely unrelated

**Comment by robbr48** [comment]
Discussion topics from regular FMI design webmeeting 2019-04-26:

**1. Disable intermediate step callbacks for temporary steps**
Providing temporary output variables for temporary solver states is only useful for specific methods such as transmission line modelling. If the FMU calls the function at every temporary state even though the master does not support it, it will induce unnecessary overhead. Hence, the master should be able to inform the FMU to not use callbacks for temporary states. It should also be possible to completely disable intermediateStepFinished callbacks.

**Proposal:** Master informs the FMU using fmi3Instantiate or fmi3SetupExperiment:
```
typedef fmi3Component fmi3InstantiateTYPE(fmi3String  instanceName,
                                          fmi3Type    fmuType,
                                          fmi3String  fmuGUID,
                                          fmi3String  fmuResourceLocation,
                                          const fmi3CallbackFunctions* functions,
                                          fmi3Boolean visible,
                                          fmi3Boolean loggingOn,
                                          fmi3Boolean intermediateVariableAccessOn,
                                          fmi3Boolean temporaryIntermediateStepsOn);
```

**2. Should an FMU be allowed to go back in time when providing "finished" output variables?**
Most (all?) solvers will never go back in time after a successful integration step. Should we restrict intermediate variable access for finished steps, so that it can only be used for points ahead of the previous finished step point? 

**Pros:** 
- Strictly incrasing time will make it easier to write master algorithms

**Cons:**
- ?

**Suggestions:**
- The name *finished integration step* is unclear should be renamed
-- *Definite point*?
-- *Finished point*?
- Enforce continously increasing time for intermediate step callbacks with step finished flag
-- Add to specification in IVA section

**Comment by masoud-najafi** [comment]
Since the callback fmi3IntermediateStepFinished()  is used for different purposes by the FMU-3.0/CS, I prepared a short document listing several usages of the callback. It may be completed, if I have missed something.
 =============================
The callback function `fmi3IntermediateStepFinished()` has been introduced in FMI-3.0 for FMU for Co-simulation. This callback function is invoked by the FMU during the execution of fmi3DoStep(). The callback is called for several purposes by the FMU.

[cols="1,2,3"]
|===
| Usage

| comment

| optional or mandatory


|Try Steps / Temporary Intermediate Steps 
|Exchange time/input/outputs at FMU internal solver minor time steps.
|Optional.

|Mesh-Points / Finished steps
|Exchange time/input/outputs at FMU internal solver mesh-points or major time steps.
|Optional.

|Inferred Clocks
|Report Inferred clock ticks happened in fmi3DoStep() to the master.
|Optional, if FMU does not support clocks.

|Inferred Events
|Report internal events happened in fmi3DoStep() such as zero-crossing events to the master.
|Optional, if FMU does not support early return.

|Discard
|Report the LastSuccsfulTime after a Discard happened in fmi3DoStep() to the master.
|Mandatory? [in FMI2.0 this functionality is avilable, so this feature may not be optional]

|Early return request
|During the callback, the cosimulation master  for time synchronisation with other FMUs or user can ask for an early return to terminate the simulation.
|Optional

|===


 The FMU may provide some of above features and the master may be interested in some or none of them. If FMU or the master does not use above features, it makes no point to define this callback.
 In this case, FMI-3.0 looks like FMI-2.0.
 
 The callback usages that are supported by the FMU are declared in the XML.  Then the master idicates the usages that are used in a CoSimulation by an API. If none of them is used, callback is not called by the FMU or the master does not define it at all. The user should be able to turn on/off the features that needs. The callabck should not be called for all of the supoorted features.  This is very similar to the way log-categories are defined for the FMI-2.0 logger.

**Comment by APillekeit** [comment]
Please replace "Inferred Clock" with "Triggered Clock" in :
|Inferred Clocks
 |Report Inferred clock ticks happened in fmi3DoStep() to the master.
 |Optional, if FMU does not support clocks.

To my knowledge, we have not used the term "inferred event" yet. We just used event. But I agree that it would be good to find a better name. Which should also be used for similar events in FMI for Model Exchange.

In my understanding an early return request is not mainly used to "terminate" the simulation but to halt the simulation at a simulation time instant for echanging values and to continue the simulation from that time instant.

**Comment by APillekeit** [comment]
I just noticed the question:
2. Should an FMU be allowed to go back in time when providing "finished" output variables?
I my view: 
As soon as the FMU has provided an intermediateStepTime with "internalStepFinished" equals True it is not allowed for the FMU to provide an earlier intermediateStepTime. If that is required the current step would have to be discarded by the FMU and the master may retry the last step for that FMU.

**Comment by robbr48** [comment]
Decision at FMI design meeting in Renningen 2019:
- `fmi3IntermediateStepFinished` is renamed to `fmi3IntermediateUpdate`
- `internalStepFinished` is renamed to `intermediateStepFinished`
- It shall not be allowed to call `fmi3IntermediateUpdate` prior to a previous call with `intermediateStepFinished=true`

**Comment by robbr48** [comment]
>     * It shall not be allowed to call `fmi3IntermediateUpdate` prior to a previous call with `intermediateStepFinished=true`

This still needs to be added to the IVA section in the specification.

**Comment by andreas-junghanns** [comment]
Someone needs to create PR!

**Comment by andreas-junghanns** [comment]
@robbr48 : Any news here?
